app/models/book.rb

```diff
@@ -6,4 +6,61 @@ class Book < ActiveRecord::Base
   has_many :favorites
   has_many :wishlists
   has_many :ratings
+  validates :title, :author, :genre, :pub_date, presence: true
+
+  scope :filter_by_title, lambda { |title|
+    where('lower(title) LIKE ?', "%#{title.downcase}%")
+  }
+
+  scope :filter_by_author, lambda { |author|
+    where('lower(author) LIKE ?', "%#{author.downcase}%")
+  }
+
+  scope :filter_by_genre, lambda { |genre|
+    where('lower(genre) LIKE ?', "%#{genre.downcase}%")
+  }
+
+  scope :filter_by_year, lambda { |year|
+    where('extract(year from pub_date) = ?', year.to_s)
+  }
+
+  scope :below_or_equal_to_avg_score, lambda { |avg_score|
+    where('avg_score <= ?', avg_score)
+  }
+
+  scope :above_or_equal_to_avg_score, lambda { |avg_score|
+    where('avg_score >= ?', avg_score)
+  }
+
+  def calculate_avg_score
+    sum = ratings.inject(0.0) { |accum, rating| accum + rating.score }
+    sum / ratings.size
+  end
+
+  def self.search(params = {})
+    books = Book.all
+    books = words_filter(books, params)
+    books = books.filter_by_year(params[:year]) if params[:year]
+    books = score_filter(books, params)
+    books
+  end
+
+  def self.words_filter(books, params = {})
+    books = books.filter_by_title(params[:title]) if params[:title]
+    books = books.filter_by_author(params[:author]) if params[:author]
+    books = books.filter_by_genre(params[:genre]) if params[:genre]
+    books
+  end
+
+  def self.score_filter(books, params = {})
+    if params[:max_score]
+      books = books.below_or_equal_to_avg_score(params[:max_score].to_f)
```

Comment:
use guard clause instead of this form of if, like this books = `books.below_or_equal_to_avg_score(params[:max_score].to_f) if params[:max_score]`